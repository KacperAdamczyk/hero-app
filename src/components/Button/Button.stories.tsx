import React from 'react';
import { action } from '@storybook/addon-actions';
import { text, select } from '@storybook/addon-knobs';
import AddIcon from '@material-ui/icons/Add';

import { Button, ButtonColor, ButtonVariant } from './Button';

export default {
  title: 'Button',
  component: Button,
};

export const Default = () => {
  const label = text('label', 'Simple button');
  const color = select(
    'color',
    [ButtonColor.primary, ButtonColor.secondary],
    ButtonColor.primary,
  );
  const variant = select(
    'variant',
    [ButtonVariant.contained, ButtonVariant.outlined],
    ButtonVariant.contained,
  );

  return (
    <Button color={color} variant={variant} onClick={action('onClick')}>
      {label}
    </Button>
  );
};

export const withIcon = () => (
  <Button
    onClick={action('onClick')}
    Icon={AddIcon}
    variant={ButtonVariant.contained}
  >
    I have an icon!
  </Button>
);

export const withLoader = () => (
  <Button onClick={action('onClick')} variant={ButtonVariant.contained} loading>
    I am loading!
  </Button>
);
