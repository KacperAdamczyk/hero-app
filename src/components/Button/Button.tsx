/** @jsx jsx */
import { FC } from 'react';
import { jsx, css } from '@emotion/core';
import UIButton from '@material-ui/core/Button';

import { colors } from 'styling';

export enum ButtonColor {
  primary = 'primary',
  secondary = 'secondary',
}

export enum ButtonVariant {
  contained = 'contained',
  outlined = 'outlined',
}

interface Props {
  color?: ButtonColor;
  variant?: ButtonVariant;
  onClick?: () => void;
  Icon?: FC;
}

const colorMap = {
  [ButtonColor.primary]: colors.primary,
  [ButtonColor.secondary]: colors.secondary,
};

export const Button: FC<Props> = ({
  children,
  color = ButtonColor.primary,
  variant = ButtonVariant.contained,
  onClick,
  Icon,
}) => (
  <UIButton
    css={css`
      && {
        text-transform: none;
        color: ${color === ButtonColor.primary ? 'white' : colorMap[color]};
        font-weight: bold;
        height: 42px;
      }
    `}
    color={color}
    variant={variant}
    onClick={onClick}
    startIcon={!!Icon && <Icon />}
  >
    {children}
  </UIButton>
);
