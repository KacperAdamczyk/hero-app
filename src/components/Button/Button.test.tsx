import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import { Button } from './Button';

describe('Button', () => {
  test('should render text', () => {
    const { getByText, container } = render(<Button>foo</Button>);

    expect(getByText('foo')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should fire onClick when clicked', async () => {
    const onClick = jest.fn();

    const { getByRole } = render(<Button onClick={onClick}>bar</Button>);

    const button = getByRole('button');
    act(() => {
      fireEvent.click(button);
    });

    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
