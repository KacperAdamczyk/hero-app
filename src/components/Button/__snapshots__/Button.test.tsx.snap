// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render text 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text css-qtic62-Button MuiButton-textPrimary"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    foo
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;
